### PR TITLE
ERRAI-1000: User is displayed as ANONYMOUS if his/her role name contains apostrophe

### PR DESCRIPTION
**Pull request by karreiro**
See https://issues.jboss.org/browse/ERRAI-1000

### CODE REVIEW

**Comment by manstis** [comment]
@karreiro I will leave review to Max however was intrigued by the JIRA. 

[Apparantly](http://www.json.org/) single quotes *should* be permitted in JSON - as ```String``` values should be enclosed in double-quotes. So it's plausible your fix patches the symptom but not the cause and possibly ```ServerMarshalling.toJSON``` may be returning incorrectly formatted JSON in the first place.

**Comment by karreiro** [comment]
@manstis thanks for your comment :-)

Since we have the JSON enclosed in one a single quote String [here](https://github.com/errai/errai/blob/master/errai-security/errai-security-server/src/main/java/org/jboss/errai/security/server/servlet/UserHostPageFilter.java#L91-L92 ), I think that would make sense escaping the ' character in the `UserHostPageFilter` class.

But, let's wait for @mbarkley opinion.

**Comment by manstis** [comment]
@karreiro ah.. perhaps use of the single quote where you linked is the issue; and it should be a double-quote with no need to then escape the single..

**Comment by mbarkley** [comment]
Hi @karreiro,

I think @manstis is correct, and that we really ought to just change the single quotes around JSON value in the filter to double quotes.

**Comment by karreiro** [comment]
Thanks @manstis and @mbarkley for the review.

I think it's done :-)

File: errai-security/errai-security-server/src/main/java/org/jboss/errai/security/server/servlet/UserHostPageFilter.java [hunk 1]
```diff
@@ -98,13 +98,26 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
   }
 
+  String securityContextJson(final User user) {
+    final String userJson = escapedJson(user);
+
+    return "{\"" + SecurityConstants.DICTIONARY_USER + "\": \"" + userJson + "\"}";
+  }
+
+  String escapedJson(final User user) {
+    final LookupTranslator translator = new LookupTranslator(new String[][]{{"\"", "\\\""}});
```
Comment (manstis) [inline errai-security/errai-security-server/src/main/java/org/jboss/errai/security/server/servlet/UserHostPageFilter.java]: It looks like you're now also protecting against use of double-quotes in the User's Roles/Groups; however should you not also then be protecting against \ too? Also, isn't a ```String.replaceAll(...)``` sufficient? Does ```LookupTranslator``` really bring anything new to the table!!?

File: errai-security/errai-security-server/src/test/java/org/jboss/errai/security/server/servlet/UserHostPageFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertTrue;
+
+public class UserHostPageFilterTest {
```
Comment (manstis) [inline errai-security/errai-security-server/src/test/java/org/jboss/errai/security/server/servlet/UserHostPageFilterTest.java]: You explicitly added support for double-quotes in the Users' Role/Groups but fail to include a test for it.

**Comment by mbarkley** [comment]
Hi @karreiro,

Sorry to belabour this whole process, but it now occurs to me that we shouldn't be wrapping the Errai marshalling JSON in any quotes at all. The value returned by `ServerMarhsalling.toJson` is already valid JSON. The only reason we were having any problems is because it was wrapped in quotes, and we should be able to get rid of all the escaping by simply not wrapping the value in quotes.

I amended your commit ([see here](https://github.com/mbarkley/errai/commit/64ca68a8cfd006717413242f4a0295034206f15d)) and it seems to work. I think this is the cleanest solution and would prefer to merge that change.

**Comment by karreiro** [comment]
Hi @mbarkley,

Nice approach, it _really is_ the cleanest. However, I think that we would need more changes to remove the double quotes. I got this error when I tested [the amended commit](https://github.com/mbarkley/errai/commit/64ca68a8cfd006717413242f4a0295034206f15d):
error

The JavaScript object `errai_security_context` is being set, but we got a problem with the IOC container.

**Comment by mbarkley** [comment]
@karreiro Good catch. I forgot to change the client-side storage handler, which was obviously a problem since it expected to parse the user from a String.

I've added another commit to [this branch](https://github.com/mbarkley/errai/tree/ERRAI-1000-master) that updates the storage handler. It required a bit more work since we did not have a marshalling API for demarshalling a JSON object directly, but I think it is worth it to not have to worry about escaping still.

WDYT?

**Comment by karreiro** [comment]
@mbarkley Perfect! :-) Do you prefer to close this PR and open another one with your branch or Can I cherry-pick your commit?

**Comment by mbarkley** [comment]
@karreiro I've pushed the commits directly:
3f346a39192ce646bbdef1a0b9ed1d2b085a5053
4626524bc4066a8b4bab12ea9bb90690b7bf7b83

**Pull request closed without merging**